src/Calculator.js

```diff
@@ -0,0 +1,191 @@
+import React from 'react';
+import { ERROR_MESSAGE, STORAGE_KEY } from './constants';
+
+class Calculator extends React.Component {
+  constructor(props) {
+    super(props);
+
+    const result = localStorage.getItem(STORAGE_KEY)
+      ? JSON.parse(localStorage.getItem(STORAGE_KEY))
+      : {
+          firstNumber: '',
+          secondNumber: '',
+          operand: '',
+        };
+
+    this.state = result;
+  }
+
+  handleNumber(e) {
+    if (this.state.firstNumber === ERROR_MESSAGE) this.clearResult();
+    if (this.state.operand === '') {
+      this.setFirstNumber(e.target.dataset.number);
+      return;
+    }
+    this.setSecondNumber(e.target.dataset.number);
+  }
+
+  handleOperation(e) {
+    if (this.state.firstNumber === ERROR_MESSAGE) this.clearResult();
+
+    if (e.target.dataset.operator === '=') {
+      this.calculate();
+      return;
+    }
+
+    if (this.state.operand !== '') return;
+    this.setState({
+      operand: e.target.dataset.operator,
+    });
+  }
+
+  setFirstNumber(value) {
+    if (this.isOverThreeDigit(this.state.firstNumber)) {
+      return;
+    }
+    this.setState((prevState) => ({
+      firstNumber: prevState.firstNumber + value,
+    }));
+  }
+
+  setSecondNumber(value) {
+    if (this.isOverThreeDigit(this.state.secondNumber)) {
+      return;
+    }
+    this.setState((prevState) => ({
+      secondNumber: prevState.secondNumber + value,
+    }));
+  }
+
+  isOverThreeDigit(number) {
+    return number.length >= 3;
+  }
+
+  calculate() {
+    if (!this.operation[this.state.operand]) return;
+
+    const result = this.operation[this.state.operand](
+      +this.state.firstNumber,
+      +this.state.secondNumber,
+    );
+
+    this.setState({
+      firstNumber: Number.isFinite(result) ? String(result) : ERROR_MESSAGE,
+      secondNumber: '',
+      operand: '',
+    });
+  }
+
+  operation = {
+    '+': this.add,
+    '-': this.subtract,
+    'x': this.multiply,
+    '/': this.divide,
+  };
+
+  add(firstNumber, secondNumber) {
+    return firstNumber + secondNumber;
+  }
+
+  subtract(firstNumber, secondNumber) {
+    return firstNumber - secondNumber;
+  }
+
+  multiply(firstNumber, secondNumber) {
+    return firstNumber * secondNumber;
+  }
+
+  divide(firstNumber, secondNumber) {
+    return Math.floor(firstNumber / secondNumber);
+  }
+
+  clearResult() {
+    this.setState({
+      firstNumber: '',
+      secondNumber: '',
+      operand: '',
+    });
+  }
+
+  saveResult(e) {
+    e.preventDefault();
+    e.returnValue = '';
+    localStorage.setItem(STORAGE_KEY, JSON.stringify({ ...this.state }));
+  }
+
+  componentDidMount() {
+    window.addEventListener('beforeunload', this.saveResult.bind(this));
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener('beforeunload', this.saveResult.bind(this));
+  }
```

Comment:
window 에 리스너를 달지 않고 그냥 componentWillUnmount 에서 saveResult 를 실행하면 되지 않을까요?
```suggestion
  componentWillUnmount() {
    this.saveResult();
  }
```

Reply:
페이지를 새로고침 하거나 탭을 꺼서 이탈할 때, 
componentWillUnmount가 호출이 안 되어서 beforeunload 이벤트 핸들러를 주는 방식으로 작성해봤는데요.
(실제로 componentWillUnmount에 localStorage save 로직을 넣고 새로고침이나 탭을 꺼버리면 저장이 안 됩니다 ㅠㅠ) 혹시 제가 잘못 알고 있는 걸까요?

Reply:
아 맞네요! 페이지 이탈 시의 코드라는걸 간과했네요 😂